Discovery strategy is still mysterious w/PEP 420 namespace packages

I'm having trouble understanding some of the subtleties of how mypy 0.701 (w/Python 3.7) discovers files to check.  Here is a sample directory tree:

```
.
└── lipy-foo
    ├── setup.cfg
    └── src
        └── linkedin
            └── foo
                ├── __init__.py
                └── bar.py
```

`linkedin` is a PEP 420 namespace package and `linkedin/foo/__init__.py` is empty.  `bar.py` contains:

```
def bar(a: int, b: int) -> str:
    return a + b
```

When I cd into the `lipy-foo` directory containing the `setup.cfg`, and try to run mypy, it can't find any `.py` files to check:

```
% ls
setup.cfg	src/
% mypy src --namespace-packages -v
There are no .py[i] files in directory 'src'
```

Clearly mypy doesn't recurse into the `src` directory to find the `linkedin` namespace package.  Other options that don't work include:

```
% mypy -p linkedin --namespace-packages
Can't find package 'linkedin'
% MYPYPATH=src mypy -p linkedin --namespace-packages
Can't find package 'linkedin'
```

What does work:

```
% mypy src/linkedin --namespace-packages
src/linkedin/foo/bar.py:2: error: Incompatible return value type (got "int", expected "str")
```

Also, if I add an empty `src/linkedin/__init__.py` file, discovery *does* work as expected:

```
% tree
.
├── setup.cfg
└── src
    └── linkedin
        ├── __init__.py
        └── foo
            ├── __init__.py
            └── bar.py
% mypy src --namespace-packages --no-site-packages
src/linkedin/foo/bar.py:2: error: Incompatible return value type (got "int", expected "str")
```

So PEP 420 style namespace packages do seem to influence the behavior here.

Since we may have additional directories and namespace packages under `src` in other repositories, I'd prefer not to have to invoke mypy with `src/linkedin`, but there doesn't seem to be any other option.

Maybe I'm doing something wrong, so please tell me if this is pebkac, or is unsupported.  *Should* mypy be able to discover the `linkedin` namespace package using `src`-style package layouts with PEP 420 namespaces?

> Would you be interested in giving this a try?

Yes!  I think it would be fine if the second invocation form was a little inconsistent.

Here's how I'm thinking about it.  Let's say I do this:

```
% touch src/linkedin/__init__.pyi
% MYPYPATH=src mypy -p linkedin --namespace-packages --no-site-packages
src/linkedin/foo/bar.py:2: error: Incompatible return value type (got "int", expected "str")
```

That makes sense to me; I want to check the `linkedin` namespace anchored underneath `src`.  What bothers me is that I have to create a dummy `__init__.pyi` file to make this work, and that this doesn't work the same way:

```
% rm -f src/linkedin/__init__.pyi 
% MYPYPATH=src mypy -p linkedin --namespace-packages --no-site-packages
Can't find package 'linkedin'
```

Is that a reasonable thing to do?

--- a/mypy/main.py
+++ b/mypy/main.py
@@ -763,7 +763,7 @@ def process_options(args: List[str],
         search_paths = SearchPaths((os.getcwd(),), tuple(mypy_path() + options.mypy_path), (), ())
         targets = []
         # TODO: use the same cache that the BuildManager will
-        cache = FindModuleCache(search_paths, fscache)
+        cache = FindModuleCache(search_paths, fscache, options, special_opts.packages)
         for p in special_opts.packages:
             if os.sep in p or os.altsep and os.altsep in p:
                 fail("Package name '{}' cannot have a slash in it.".format(p),

--- a/mypy/modulefinder.py
+++ b/mypy/modulefinder.py
@@ -64,7 +64,8 @@ class FindModuleCache:
     def __init__(self,
                  search_paths: SearchPaths,
                  fscache: Optional[FileSystemCache] = None,
-                 options: Optional[Options] = None) -> None:
+                 options: Optional[Options] = None,
+                 ns_packages: Optional[List[str]] = None) -> None:
         self.search_paths = search_paths
         self.fscache = fscache or FileSystemCache()
         # Cache for get_toplevel_possibilities:
@@ -74,6 +75,7 @@ class FindModuleCache:
         self.results = {}  # type: Dict[str, Optional[str]]
         self.ns_ancestors = {}  # type: Dict[str, str]
         self.options = options
+        self.ns_packages = ns_packages or []  # type: List[str]
 
     def clear(self) -> None:
         self.results.clear()
@@ -307,6 +309,13 @@ class FindModuleCache:
                     if mod not in hits:
                         hits.add(mod)
                         result += self.find_modules_recursive(module + '.' + mod)
+        elif os.path.isdir(module_path) and module in self.ns_packages:
+            # Even more subtler: handle recursive decent into PEP 420
+            # namespace packages that are explicitly listed on the command
+            # line with -p/--packages.
+            for item in sorted(self.fscache.listdir(module_path)):
+                if os.path.isdir(os.path.join(module_path, item)):
+                    result += self.find_modules_recursive(module + '.' + item)
         return result
 
 

--- a/test-data/unit/cmdline.test
+++ b/test-data/unit/cmdline.test
@@ -1124,6 +1124,19 @@ p/a.py:4: error: Argument 1 to "foo" has incompatible type "str"; expected "int"
 p/b/__init__.py:5: error: Argument 1 to "bar" has incompatible type "str"; expected "int"
 c.py:2: error: Argument 1 to "bar" has incompatible type "str"; expected "int"
 
+[case testSrcPEP420Packages]
+# cmd: mypy -p anamespace --namespace-packages
+[file mypy.ini]
+[[mypy]]
+mypy_path = src
+[file src/setup.cfg]
+[file src/anamespace/foo/__init__.py]
+[file src/anamespace/foo/bar.py]
+def bar(a: int, b: int) -> str:
+    return a + b
+[out]
+src/anamespace/foo/bar.py:2: error: Incompatible return value type (got "int", expected "str")
+
 [case testFollowImportStubs1]
 # cmd: mypy main.py
 [file mypy.ini]

